p4-16/spec/P4-16-spec.mdk

```diff
@@ -1639,6 +1642,60 @@ directions:
   including values for the directionless parameters. The directionless
   parameters in this case behave like `in` parameters.
   See Section [#sec-invoke-actions] for further details.
+- Default parameter values are only allowed for 'in' or direction-less parameters;
+  these values must evaluate to compile-time constants.
+
+### Optional parameters
+
+A parameter that is annotated with the `@optional` annotation is
+optional: the user may supply no value for that parameter in an
+invocation.  Optional parameters can only appear for arguments of:
+packages, extern functions, extern methods, and extern object
+constructors.  Optional parameters cannot have default values.  If a
+function type has both optional parameters and default values then it
+can only be called using named arguments.  It is recommended, but not
+mandatory, for all optional parameters to be at the end of a parameter
+list.
+
+The implementation of such objects is not expressed in P4, so the
+meaning and implementation of optional parameters should be specified
+by the target architecture.  For example, we can imagine a two-stage
+switch architecture where the second stage is optional.  This could be
+declared as a package with an optional parameter:
+
+~Begin P4Example
+package pipeline(...);
+package switch(pipeline first, @optional pipeline second);
```

Comment:
In this example, if both pipelines are optional: 
`package switch(@optional pipeline first, @optional pipeline second);` 

and if user writes:
`pipeline (...) egress;`
`switch(egress) main;`

There is an ambiguity on which pipeline argument is defined by the program. The spec did not say anything about using the named argument to disambiguate.

Reply:
Overload resolution will not be very easy to implement. 
Also support for default parameter values is not yet implemented.

Reply:
If overload resolution is implemented, do we still need the `optional` annotation? If I understand correctly, the original decision to have `@optional` was to close the gap while overload resolution is being implemented. But now we are promoting optional annotation to the language. We will end up with two mechanisms to implement the variable arguments feature.

Reply:
Overloads are useful even if we don't have optional arguments.
Also, n optional arguments can express 2^n overloads, so they are still useful.